compiler/rustc_pattern_analysis/src/rustc.rs

```diff
@@ -910,6 +907,22 @@ impl<'p, 'tcx: 'p> RustcPatCtxt<'p, 'tcx> {
     }
 }
 
+/// Returns `true` if the given pattern would be printed as a wildcard (`_`).
+fn would_print_as_wildcard(p: &WitnessPat<'_, '_>) -> bool {
+    match p.ctor() {
+        Constructor::IntRange(IntRange {
+            lo: MaybeInfiniteInt::NegInfinity,
+            hi: MaybeInfiniteInt::PosInfinity,
+        })
+        | Constructor::Never
```

Comment:
`!` should not count as a wildcard here, since `[!, ..]` is semantically different from `[..]`.

```suggestion
```

Reply:
If you really want precision, you can do
```suggestion
        | Constructor::Never if tcx.features().never_patterns
```

Reply:
I'll stick with the less-precise version for now, because requiring `tcx` is going to be somewhat awkward (temporarily) in an upcoming change.

Is it possible for `Constructor::Never` patterns to exist when the feature is disabled? It wasn't obvious to me why the print behaviour is conditional.

Reply:
Never mind; turns out my future changes will need the precise version to avoid breaking existing tests, and supplying `tcx` is actually fine.